### PR TITLE
refactor: change BaseClass.baseInterface from length/ptr to [ ]

### PR DESCRIPTION
**Pull request by WalterBright**
Could go further with this, but smaller PRs are better.

### CODE REVIEW

**Comment by yebblies** [comment]
Why not just use `Array`?

**Comment by WalterBright** [comment]
It isn't allocated like Array.

**Comment by yebblies** [comment]
I don't see any use that would prevent using `Array`... Why shouldn't it?  That would avoid calling xcalloc directly too.

File: src/dtemplate.d [hunk 1]
```diff
@@ -3928,9 +3928,10 @@ extern (C++) MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplatePara
                 }
             }
             // Now recursively test the inherited interfaces
-            for (size_t j = 0; j < b.baseInterfaces_dim; ++j)
+            //for (size_t j = 0; j < b.baseInterfaces_dim; ++j)
```
Comment (yebblies) [inline src/dtemplate.d]: Dead code.

**Comment by WalterBright** [comment]
Array is a C++ clone of D arrays. The only reason to keep using them is to interface with the C++ code. baseInterfaces[ ] does not need to interface to the C++ code. It's also fixed size (doesn't need any resizing logic).

**Comment by ibuclaw** [comment]
I could be a user.  Let me check quickly.

**Comment by ibuclaw** [comment]
Nope, ``baseInterface`` is only used in the frontend.

File: src/dclass.d [hunk 2]
```diff
@@ -1850,7 +1849,7 @@ public:
     bool isBaseOf(BaseClass* bc, int* poffset)
     {
         //printf("%s.InterfaceDeclaration::isBaseOf(bc = '%s')\n", toChars(), bc->sym->toChars());
-        for (size_t j = 0; j < bc.baseInterfaces_dim; j++)
+        for (size_t j = 0; j < bc.baseInterfaces.length; j++)
```
Comment (JackStouffer) [inline src/dclass.d]: `foreach(ref b; bc.baseInterfaces)`?

**Comment by MartinNowak** [comment]
> It isn't allocated like Array.

It is, root.Array (now?) contains value types. Would make sense to use it here.

**Comment by WalterBright** [comment]
> Would make sense to use it here.

No. It makes no sense to reinvent D's array type here. `Array` is not a better array type. It's not a Phobos type, either. It exists purely because the original was in C++ and didn't have D arrays available.

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
As usual, Andrei addresses everyone's objections.

**Comment by ibuclaw** [comment]
> As usual, Andrei addresses everyone's objections.

"Turn him to any cause of policy, the Gordian Knot of it he will unloose, familiar as his garter"

**Pull request merged**